### ISSUE TITLE
Native Wayland support

### ISSUE THREAD

**Issue by iMonZ** [shundhammer/qdirstat#153]
The app: "QdirStat" doesn't support native Wayland.

I get this error while starting the application:

```
[2021-02-19 15:24:16.497 [387423] <Info>    Logger.cpp:143 openLogFile():  -- Log Start --
2021-02-19 15:24:16.497 [387423] <Info>    main.cpp:58 logVersion():  QDirStat-1.7 built with Qt 5.15.1
2021-02-19 15:24:16.497 [387423] <Info>    [Qt] Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway.
2021-02-19 15:24:16.498 [387423] <WARNING> [Qt] Could not find the Qt platform plugin "wayland" in ""
2021-02-19 15:24:16.498 [387423] <ERROR>   [Qt] This application failed to start because no Qt platform plugin could be initialized. 
2021-02-19 15:24:16.498 [387423] <ERROR>   [Qt] Available platform plugins are: eglfs, linuxfb, minimal, minimalegl, offscreen, vnc, xcb.
2021-02-19 15:24:16.498 [387423] <ERROR>   Logger.cpp:406 qt_logger():  FATAL: Could not connect to the display.
2021-02-19 15:24:16.498 [387423] <Info>    Logger.cpp:413 qt_logger():  -- Exiting --]
```

Please fix this with adding the Wayland support (QT already support native Wayland)

**Comment by shundhammer** [comment]
QDirStat supports everything that Qt supports. What makes you think it doesn't?
If it doesn't start on Wayland, the problem is on your desktop's side, not on QDirStat's.

And the message (coming from deep inside the Qt libs) already tells you what you can try:

> Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway.

**Comment by shundhammer** [comment]
See also the lengthy explanation here:

https://github.com/shundhammer/qdirstat/blob/master/src/Logger.cpp#L362-L407

**Comment by iMonZ** [comment]
QdirStat runs with xWayland:
https://wayland.freedesktop.org/xserver.html

When you disable this.
You can do that with setting this variable: QT_QPA_PLATFORM=wayland

As I know Wayland must be added inside of QT to the compiler to fix this error

**Comment by shundhammer** [comment]
For one thing, it's QDirStat and Qt, not QdirStat and QT (which would be Apple QuickTime).

For another, did you even read that documentation that you linked to?

It's about building the QtWayland platform plug-in for Qt. That has nothing to do with building the application. The whole purpose of Qt is being a cross-platform toolkit, i.e. abstracting from that level of windowing system technicalities. It doesn't matter if a program runs on Windows, MacOS X, X11, or Wayland; it's always Qt, and Qt takes care of that stuff.

On a Linux platform, you might or might not have the QtWayland plug-in installed. On Ubuntu, install package _qtwayland5_.

And frankly, a user doesn't even need to care what hoops the system goes through to run an application, if it's using XWayland or a QtWayland plug-in.

Also, I am certainly not change QDirStat in a way that makes it incompatible with 80% of all Linux/BSD/Unix users who use X11.

**Comment by shundhammer** [comment]
## tl;dr

_**Make sure you have the QtWayland plug-in installed.**_

Ubuntu:

    sudo apt install qtwayland5

**Comment by shundhammer** [comment]
> The app: "QDirStat" doesn't support native Wayland.

Wrong; it does. Or, more exactly, **Qt does**. 


> Please fix this with adding the Wayland support 

Wrong. Qt takes care of that. That's the whole point of a multi-platform toolkit like Qt.


> As I know Wayland must be added inside of Qt to the compiler to fix this error.

Wherever this nugget of information comes from, it's plain wrong. Please do not spread computer voodoo!

**Comment by iMonZ** [comment]
Hey thank you it was my fault! (Or an bug from the operating system)
The tool is working fine now

**Comment by shundhammer** [comment]
_**Are you actively spreading this bullshit all over the Internet 3 days AFTER I explained it to you?**_

https://answers.launchpad.net/ubuntu/+question/695689

Seriously, it is okay to not know something. That's fair. People have to start somewhere, and learning about such a big topic like Linux is a life-long task.

But you should *know* that your knowledge is limited. Before you keep lecturing others, do at least some basic research. Ask knowledgeable people before you accuse othere of being too stupid. And when they take their time explaining things to you, **do not spread misinformation any further!**

You have a dangerous half-knowledge paired with huge overconfidence. You have a lot to learn; and on that long journey that you have ahead of you, listen before you talk! Stop making unsubstantiated claims!

**Comment by shundhammer** [comment]
> Hey thank you the problem is fixed now!

Seriously? I believe it never existed outside of your imagination.

> It was a fault of the Ubuntu universe repo.

I find that hard to believe. Where can I read more details about that?

**Comment by iMonZ** [comment]
> > Hey thank you the problem is fixed now!
> 
> Seriously? I believe it never existed outside of your imagination.
> 
> > It was a fault of the Ubuntu universe repo.
> 
> I find that hard to believe. Where can I read more details about that?

I’m sorry I have removed qdirstat and VLC from the bugreport.

For Telegram-Desktop and Firefox the issue is already confirmed 

Since both problems drop the same error you should understand why it is confusing

**Comment by shundhammer** [comment]
I believe you are confused how Wayland, XWayland and Qt-based programs are interacting.

The reactions in that bug report suggest that others are in turn confused about what problem you believe exists.

Wayland was (yet another) attempt to reinvent the wheel; back many years ago they believed that the performance of X11 could be vastly improved by rewriting it completely without features they considered unnecessary like network transparancy and so many things that modern high-performance software needs being moved out to optional X11 extensions.

So they did reinvent the wheel, they rewrote a new windowing system from scratch with everything being local on the same machine, no network delays, many tasks being done directly in the application (which used to be done in a distributed environment like an X server that might possibly be on a remote machine in the network).

And then they found out that they needed backwards compatibility to X11 because users wanted to run all the existing X applications as well, not just the handful of dedicated Wayland applications. XWayland was born.

Trolltech / the Qt project strongly believed that Wayland would be the future, so they added another windowing system to Qt (QtWayland) in addition to the existing X11, MS Windows, MacOS X. But they also realized that Wayland had a long time ahead until its world domination.

So, hammering the point home once more: Any Qt program does not need to do anything special to work with Wayland. It's up to the user to select the correct plug-in; while the default is still XCB (X11), you as the user can choose QtWayland if you (a) install that package and (b) set up your desktop environment to use it as the default.

Other (non-Qt) applications may choose to add their own abstraction layer to be able to use either X11 or Wayland, but if there is no such abstraction layer yet, this would basically mean a rewrite from scratch which most projects will probably not do. After all, XWayland is providing the X11 API for that exact purpose.

What does a user lose when using XWayland? Maybe some graphics performance; but for applications that are not doing much graphics anyway, just typical GUI elements like buttons and lists, there will be no noticeable difference.

What does an application lose when deciding to go all Wayland and drop X11 support? A large portion of their user base; all those users who are not using Wayland. That may be more than you expect.

What does an application lose when trying to support both, X11 and Wayland?

If they don't have any abstraction layer yet, it comes down to a complete rewrite of all the GUI parts of the application. But most applications are using some GUI toolkit anyway; but many of those are based on X11 / Xlib or one of the higher levels like Xt. So it would come down to adding Wayland support to those toolkits which will be when hell freezes over (plus 3 days). That's what XWayland is for.

So, I don't know what fight you are fighting, but IMHO it's a losing battle. Unless you want to rewrite all those X11 applications yourself, I don't think it's realistic to expect them to all add native Wayland support.

**Comment by iMonZ** [comment]
Hey thank you for your answer but currently I'm only complaining about QT apps
which you already said works with wayland!

Xwayland has a battery issue on mobile devices too so a switch to wayland is great!

Also I haven't complained about any apps that
1. aren't using QT
2. aren't using any toolkit that already support wayland

**Comment by shundhammer** [comment]
So fixing existing problems in XWayland should be the number one priority. I don't think it's realistic that XWayland will go away anytime soon.

This is very much like the old MS-DOS compatibility that exists until this day in MS Windows: Everybody knows that it's not an ideal solution, but it's there, it works, and it makes some users happy because it keeps the software running that they like to use (no matter if a newer and better equivalent is available).
